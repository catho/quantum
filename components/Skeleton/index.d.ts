import { FC } from 'react';

export interface SkeletonButtonProps {
  size?: 'xsmall' | 'small' | 'medium' | 'large';
}

export interface SkeletonProps {
  type?: 'rect' | 'circle' | 'text' | 'button' | 'tag';
  as?: 'div' | 'span';
  width?: string;
  height?: string;
  size?: 'xsmall' | 'small' | 'medium' | 'large';
}
export interface SkeletonTagProps {
  size?: 'xsmall' | 'small' | 'medium' | 'large';
}

declare const SkeletonText: FC<SkeletonProps>;
declare const SkeletonButton: FC<SkeletonButtonProps>;
declare const SkeletonTag: FC<SkeletonTagProps>;
declare const SkeletonCircle: FC;

declare const Skeleton: FC<SkeletonProps> & {
  Button: typeof SkeletonButton;
  Circle: typeof SkeletonCircle;
  Text: typeof SkeletonText;
  Tag: typeof SkeletonTag;
}

export { SkeletonText, SkeletonButton, SkeletonTag, SkeletonCircle};
export default Skeleton;
