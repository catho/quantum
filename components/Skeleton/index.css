@import "./sub-components/Button.module.css";
@import "./sub-components/Rect.module.css";
@import "./sub-components/Tag.module.css";
@import "./sub-components/Text.module.css";
@import "./sub-components/Circle.module.css";