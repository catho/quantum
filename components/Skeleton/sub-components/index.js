import SkeletonRect from './SkeletonRect';
import SkeletonCircle from './SkeletonCircle';
import SkeletonText from './SkeletonText';
import SkeletonButton from './SkeletonButton';
import SkeletonTag from './SkeletonTag';

export {
  SkeletonRect,
  SkeletonCircle,
  SkeletonText,
  SkeletonButton,
  SkeletonTag,
};
