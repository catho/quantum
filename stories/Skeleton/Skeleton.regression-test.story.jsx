import React from 'react';
import { storiesOf } from '@storybook/react';
import { Skeleton } from '../../components';

storiesOf('Skeleton', module)
  .add('rect as default', () => (
    <>
      <Skeleton />
      <br />
      <Skeleton width="100px" />
      <br />
      <Skeleton height="20px" width="100px" />
    </>
  ))
  .add('text', () => (
    <>
      <Skeleton.Text />
      <br />
      <Skeleton.Text width="100px" />
      <br />
      <Skeleton.Text height="20px" width="100px" />
    </>
  ))
  .add('circle', () => <Skeleton.Circle height="56px" width="56px" />)
  .add('Button', () => (
    <>
      <Skeleton.Button width="100px" />
      <br />
      <Skeleton.Button width="10%" />
      <br />
      <Skeleton.Button size="xsmall" />
      <br />
      <Skeleton.Button size="small" />
      <br />
      <Skeleton.Button size="medium" />
      <br />
      <Skeleton.Button size="large" />
    </>
  ))
  .add('Tag', () => (
    <>
      <Skeleton.Tag width="100px" />
      <Skeleton.Tag width="100px" />
      <br />
      <Skeleton.Tag width="10%" />
      <br />
      <Skeleton.Tag width="200px" size="small" />
      <Skeleton.Tag width="50px" size="small" />
      <br />
      <Skeleton.Tag width="200px" size="medium" />
      <Skeleton.Tag width="50px" size="medium" />
      <br />
      <Skeleton.Tag width="200px" size="large" />
      <Skeleton.Tag width="50px" size="large" />
    </>
  ));
