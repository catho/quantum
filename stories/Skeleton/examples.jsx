import React from 'react';
import { Skeleton, Card, Tag, Button } from '../../components';

export default [
  {
    code: '<Skeleton />',
    component: <Skeleton />,
  },
  {
    code: '<Skeleton />',
    component: <Skeleton.Button />,
  },
  {
    code: '',
    component: (
      <>
        <Card>
          <Card.Header>
            <Card.HeaderText>
              <Card.Title small>Administrador</Card.Title>
              <Card.Description small>
                <Tag text="Hortolândia – SP" /> <Tag text="R$ 1.200" />
              </Card.Description>
            </Card.HeaderText>
            <Card.Thumbnail rounded src="#" alt="" />
          </Card.Header>
          <Card.Content>
            Irá atuar com Atendimento e recepção de clientes para reuniões e
            eventos agendados na empresa. Experiência em atendimento e recepção
            de clientes{' '}
            <a href="#" style={{ fontWeight: '700' }}>
              continuar lendo…
            </a>
          </Card.Content>
          <Card.Footer>
            <Button size="small">Enviar currículo</Button>
          </Card.Footer>
        </Card>
        <Card>
          <Card.Header>
            <Card.HeaderText>
              <Card.Title small>Administrador</Card.Title>
              <Card.Description small>
                <Tag text="Hortolândia – SP" /> <Tag text="R$ 1.200" />
              </Card.Description>
            </Card.HeaderText>
            <Card.Thumbnail rounded src="#" alt="" />
          </Card.Header>
          <Card.Content>
            Irá atuar com Atendimento e recepção de clientes para reuniões e
            eventos agendados na empresa. Experiência em atendimento e recepção
            de clientes{' '}
            <a href="#" style={{ fontWeight: '700' }}>
              continuar lendo…
            </a>
          </Card.Content>
          <Card.Footer>
            <Button size="small">Enviar currículo</Button>
          </Card.Footer>
        </Card>
      </>
    ),
  },
];
