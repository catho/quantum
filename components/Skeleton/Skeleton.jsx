import React from 'react';
import PropTypes from 'prop-types';
import SkeletonTypes from './SkeletonTypes';
import { spacing, baseFontSize as defaultBaseFontSize } from '../shared/theme';
import {
  SkeletonRect,
  SkeletonCircle,
  SkeletonText,
  SkeletonButton,
  SkeletonTag,
} from './sub-components';

const Skeleton = props => {
  const { type, size, theme, height, width } = props;

  const rectProps = {
    height,
    width,
  };
  const circleProps = {
    height,
    width,
  };

  const textProps = {
    height,
    width,
  };

  const buttonProps = {
    size,
    theme,
    width,
  };

  const tagProps = {
    size,
    theme,
    width,
  };

  return (
    <>
      {type === 'rect' && <SkeletonRect {...rectProps} />}
      {type === 'circle' && <SkeletonCircle {...circleProps} />}
      {type === 'text' && <SkeletonText {...textProps} />}
      {type === 'button' && <SkeletonButton {...buttonProps} />}
      {type === 'tag' && <SkeletonTag {...tagProps} />}
    </>
  );
};

Skeleton.defaultProps = {
  type: 'rect',
  width: '100%',
  height: '14px',
  size: 'medium',
  theme: {
    baseFontSize: defaultBaseFontSize,
    spacing,
  },
};

Skeleton.propTypes = {
  /** Sets the component behavior */
  type: PropTypes.oneOf(['rect', 'circle', 'text', 'button', 'tag']),
  width: PropTypes.string,
  /** Will affect only types that not uses the prop `size` */
  height: PropTypes.string,
  /** Will affect only `Tag` and `Button` types.  */
  size: PropTypes.oneOf(['xsmall', 'small', 'medium', 'large']),
  /** Used only for themification.  */
  theme: PropTypes.shape({
    spacing: PropTypes.object,
    baseFontSize: PropTypes.number,
  }),
};

// Types
Skeleton.Button = SkeletonTypes.Button;
Skeleton.Circle = SkeletonTypes.Circle;
Skeleton.Text = SkeletonTypes.Text;
Skeleton.Tag = SkeletonTypes.Tag;

export default Skeleton;
