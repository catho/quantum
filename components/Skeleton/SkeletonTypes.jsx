import React from 'react';
import Skeleton from './Skeleton';

const SkeletonTypes = {
  Circle: props => <Skeleton type="circle" {...props} />,
  Text: props => <Skeleton type="text" {...props} />,
  Button: props => <Skeleton type="button" {...props} />,
  Tag: props => <Skeleton type="tag" {...props} />,
};

SkeletonTypes.Circle.displayName = 'Skeleton.Circle';
SkeletonTypes.Text.displayName = 'Skeleton.Text';
SkeletonTypes.Button.displayName = 'Skeleton.Button';
SkeletonTypes.Tag.displayName = 'Skeleton.Tag';

export default SkeletonTypes;
