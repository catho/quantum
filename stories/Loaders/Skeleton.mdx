import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import Header from '../shared/Header';

import Skeleton from '../../components/Skeleton';
import { Default, WithType, WithSize } from './Skeleton.stories.jsx';

<Meta title="Loaders/Skeleton" component={Skeleton} />

<Header title="Skeleton">
    A component that replaces others while they are loading or not ready to render.
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import Skeleton from '@catho/quantum/Skeleton';
```

<Canvas>
  <Story story={Default}/>
</Canvas>

## <a id="api"></a>API

<ArgsTable of={Skeleton} />

## <a id="examples"></a>Examples

### With type

<Canvas>
  <Story story={WithType}/>
</Canvas>

### With size

<Canvas>
  <Story story={WithSize}/>
</Canvas>

