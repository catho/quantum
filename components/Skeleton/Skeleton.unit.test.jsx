import React from 'react';
import renderer from 'react-test-renderer';
import Skeleton from './Skeleton';

describe('Skeleton component', () => {
  it('Should match the snapshot of a simple skeleton', () => {
    const simpleSkeleton = <Skeleton />;
    expect(renderer.create(simpleSkeleton).toJSON()).toMatchSnapshot();
  });

  it('Should match the snapshot of a text skeleton', () => {
    const textSkeleton = <Skeleton.Text />;
    expect(renderer.create(textSkeleton).toJSON()).toMatchSnapshot();
  });

  it('Should match the snapshot of a circular skeleton', () => {
    const circularSkeleton = <Skeleton.Circle height="56px" width="56px" />;
    expect(renderer.create(circularSkeleton).toJSON()).toMatchSnapshot();
  });

  it('Should match the snapshot of a Button component skeleton', () => {
    const buttonSkeleton = <Skeleton.Button width="100px" />;
    expect(renderer.create(buttonSkeleton).toJSON()).toMatchSnapshot();
  });

  it('Should match the snapshot of a Tag component loader', () => {
    const tagSkeleton = <Skeleton.Tag width="100px" />;
    expect(renderer.create(tagSkeleton).toJSON()).toMatchSnapshot();
  });
});
