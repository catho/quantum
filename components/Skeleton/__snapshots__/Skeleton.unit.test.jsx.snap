// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton component Should match the snapshot of a Button component skeleton 1`] = `
.c0 {
  background-color: rgba(0,0,0,0.08);
  -webkit-animation: bzcbYv 1.5s ease-in-out infinite;
  animation: bzcbYv 1.5s ease-in-out infinite;
  display: inline-block;
  box-sizing: border-box;
  width: 100px;
  border-radius: 4px;
  height: 40px;
}

<div
  className="c0"
  height={null}
  size="medium"
  width="100px"
/>
`;

exports[`Skeleton component Should match the snapshot of a Tag component loader 1`] = `
.c0 {
  background-color: rgba(0,0,0,0.08);
  -webkit-animation: bzcbYv 1.5s ease-in-out infinite;
  animation: bzcbYv 1.5s ease-in-out infinite;
  display: inline-block;
  box-sizing: border-box;
  width: 100px;
  border-radius: 8px;
  margin-right: 8px;
  height: 26px;
}

<div
  className="c0"
  height={null}
  size="medium"
  width="100px"
/>
`;

exports[`Skeleton component Should match the snapshot of a circular skeleton 1`] = `
.c0 {
  background-color: rgba(0,0,0,0.08);
  -webkit-animation: bzcbYv 1.5s ease-in-out infinite;
  animation: bzcbYv 1.5s ease-in-out infinite;
  display: inline-block;
  box-sizing: border-box;
  height: 56px;
  width: 56px;
  border-radius: 50%;
}

<div
  className="c0"
  height="56px"
  width="56px"
/>
`;

exports[`Skeleton component Should match the snapshot of a simple skeleton 1`] = `
.c0 {
  background-color: rgba(0,0,0,0.08);
  -webkit-animation: bzcbYv 1.5s ease-in-out infinite;
  animation: bzcbYv 1.5s ease-in-out infinite;
  display: inline-block;
  box-sizing: border-box;
  height: 14px;
  width: 100%;
}

<div
  className="c0"
  height="14px"
  width="100%"
/>
`;

exports[`Skeleton component Should match the snapshot of a text skeleton 1`] = `
.c0 {
  background-color: rgba(0,0,0,0.08);
  -webkit-animation: bzcbYv 1.5s ease-in-out infinite;
  animation: bzcbYv 1.5s ease-in-out infinite;
  display: inline-block;
  box-sizing: border-box;
  height: 14px;
  width: 100%;
  border-radius: 4px;
}

<div
  className="c0"
  height="14px"
  width="100%"
/>
`;
